### DISCUSSION TITLE
IP network bypassing mechanisms

### DISCUSSION THREAD

**Discussion by cperera-aud** [zephyrproject-rtos/zephyr#55455] [Q&A] [answered]
My question is, does Zephyr provide a means to bypass the IP network entirely and be able to send and receive data? e.g. Send/receive audio wrapped in UDP/IP headers directly to/from the ethernet peripheral. Just to make things a bit more complicated, would it be possible to also send/receive UDP/IP data via the network stack? So in essence one type of data designated by its UDP/IP parameters would need to bypass the stack whereas another set of data should traverse the IP stack. 

If such mechanisms exist, is it possible to point me at samples, config params, any available docs? Any help is always much appreciated!

**Comment by jukkar** [comment, marked as answer]
If you have network interface and do not want to use a packet sockets, then it is possible to call `net_send_data()` directly to send pre-formatted data. Note that you would need to build the data packet yourself with all the needed protocol layers.
For the RX side, it is more complicated and for normal IP traffic not possible. You could have your own L2 layer and then you would receive all the network packets for a specific network interface, but then you cannot use the normal sockets to receive those packets.
Of course as the source code is available, you can certainly tweak the sources as you wish, see `net_recv_data()` that is called by network device driver when a packet is received, and the functions it calls and do needed changes. This would not be upstreamable of course.

If you just want to send raw UDP/IP packets that you construct yourself, then I would recommend using the packet socket type in which case you could still use normal BSD socket interface to send/receive normal application data in the same network interface.

**Reply by cperera-aud** [reply]
Thank you @jukkar for your reply, much appreciated. I gather certain solutions would not be upstreamable. I'll have a look at the functions that you mentioned.

One question that I had was in the case of raw sockets I was trying to open up a socket to receive UDP/IP data but it didn't seem like the data was being received. Sorry I gave up after a while and thought of trying something else so I don't have the changes that I was trying. For example in the case of having an ethernet interface, would you use the following to open a raw socket: 

`socket(AF_PACKET, SOCK_RAW, ETH_P_ALL)`

Do you know off hand if there was a sample that made use of a raw sockets to receive UDP/IP?

**Reply by jukkar** [reply]
See for example `samples/net/sockets/packet` for an example of raw socket usage.

**Reply by cperera-aud** [reply]
Thank you again @jukkar, much appreciated for your help.
